Adding browser check doc (#3901)
DataDog/documentation: content/synthetics/browser_check.md

    {{< img src="synthetics/browser_check/browser_check_assertions.png" alt="Browser Check assertions" responsive="true" style="width:80%;">}}

    **Note**: **Your last browser check step must be an Assertion**, otherwise there is nothing to check.
5. Once you have finished your Scenario, click on **Save and Launch Test**.

### Actions
#### Assertion

{{< img src="synthetics/browser_check/assertions_browser_check.png" alt="Assertion browser check" responsive="true" style="width:40%;">}}

Assertions allow you to check if an element, a content, or some text is available in the current page:

| Assertion                                               | Description                                                                                                                      |
| ----                                                    | ----                                                                                                                             |
| `Assert that an element is present on the page`         | Asserts that an element (such as a specific `span`, `div`, `h`, `a`, etc.) is present on the current page.                       |

> | `Check an element's content`                            | Makes sure that a specific element is located or not on the current page.                                                        |

Reviewer: Text a bit misleading, since it's more about checking the *content* (= text) of an element, not the presence of the element, even though of course if the element is absent its content cannot be checked and so the assertion will fail.

Author: let's wait for some customer feedback on this one.

| `Assert that some text is present anywhere on the page` | Asserts that some specific text is present on the current page.                                                                  |
| `Assert that some text is nowhere on the page`          | Asserts that some specific text is **NOT** present on the current page.                                                          |
| `Check main page URL's content`                         | This takes the URL of the current page, and asserts whether a specific value (`string`, `number`, `regex`) is present within it. |

#### Navigation

The navigation action allow you to: 

* Refresh the current page of the scenario.
* Follow a specific link.

#### Hover

This browser check step isn’t added through an actual hovering mechanism (otherwise each element you are hovering would be added as a step) but using a dedicated action with a click.

> After selecting the Hover action, the hovering is blocked, and the step can be added by clicking on the element that you want to choose. 

Reviewer: Not sure what `the hovering is blocked` means, maybe something like `After selecting the Hover action, click on the element you want to choose and a step will be created.`

1. Optionally, select **Open in a pop-up** at the upper right of the page to open your test recording in a separate pop-up window in order to avoid sizing issues in the displayed window within Datadog's interface.
2. Click on **Start recording** to begin recording your Browser Check.
3. Your actions are recorded and used to create steps within your Browser Check Scenario.
4. Use the actions available in the upper left corner to enrich your scenario:
    {{< img src="synthetics/browser_check/browser_check_assertions.png" alt="Browser Check assertions" responsive="true" style="width:80%;">}}

    **Note**: **Your last browser check step must be an Assertion**, otherwise there is nothing to check.
5. Once you have finished your Scenario, click on **Save and Launch Test**.

### Actions
#### Assertion

{{< img src="synthetics/browser_check/assertions_browser_check.png" alt="Assertion browser check" responsive="true" style="width:40%;">}}

Assertions allow you to check if an element, a content, or some text is available in the current page:

| Assertion | Description |
| ----      | ----        |

> | `Assert that an element is present on the page` | Asserts that an element is present on the page. This checks if a specific `span`, `div`, `h`, `a`, etc. is somewhere on the current page. |

Reviewer 2: ```suggestion
| `Assert that an element is present on the page` | Asserts that an element (such as a specific `span`, `div`, `h`, `a`, etc.) is present on the current page. |
```

> | `Check an element's content` |  Makes sure that a specific element is located or not on the current page. |

Reviewer 2: From the description here, it seems like this isn't different from "Assert that an element is present on the page." Perhaps it should be "Asserts that the content of a specific element is present on the current page."

> | `Assert that some text is present anywhere on the page` | Makes sure that a specific text is located on the current page. |

Reviewer 2: ```suggestion
| `Assert that some text is present anywhere on the page` | Asserts that some specific text is present on the current page. |
```

> | `Assert that some text is nowhere on the page` | Makes sure that a specific text is **NOT** located on the current page. |

Reviewer 2: ```suggestion
| `Assert that some text is nowhere on the page` | Asserts that some specific text is **NOT** present on the current page. |
```

> | `Check main page URL's content` | This takes the URL or the current page, and check whether or not a specific value (`string`, `number`, `regex`) is located within it. |

Reviewer 2: ```suggestion
| `Check main page URL's content` | This takes the URL of the current page, and asserts whether a specific value (`string`, `number`, `regex`) is present within it. |
```

#### Navigation

The navigation action allow you to: 

* Refresh the current page of the scenario.
* Follow a specific link.

#### Hover

This browser check step isn’t added through an actual hovering mechanism (otherwise each element you are hovering would be added as a step) but using a dedicated action with a click.

After selecting the Hover action, the hovering is blocked, and the step can be added by clicking on the element that you want to choose. 

#### Variable

##### Define a variable

{{< img src="synthetics/browser_check/setup_variable.png" alt="Setup Variable" responsive="true" style="width:40%;">}}

> To define a variable, enter a capitalized name then choose its type between 

Reviewer 2: ```suggestion
To define a variable, enter a capitalized name then choose its type between:
```

> * `Select an element content` : this type allows you to create a variable out of a `span`, `div`, etc. content by extracting the text of this element.

Reviewer 2: ```suggestion
* `Select an element content`: this type allows you to create a variable out of a `span`, `div`, etc. content by extracting the text of this element.
```

---
title: Browser Check
kind: documentation
beta: true
description: Simulate and monitor user journeys from specific locations.
further_reading:
- link: "https://www.datadoghq.com/blog/introducing-synthetic-monitoring/"
  tag: "Blog"
  text: "Introducing Datadog Synthetics"
- link: "synthetics/"
  tag: "Documentation"
  text: "Manage your checks"
- link: "synthetics/uptime_check"
  tag: "Documentation"
  text: "Configure an Uptime Check"
---

> <div class="alert alert-warning">Synthetics is in beta for the Datadog US Site. Request access via the <a href="https://app.datadoghq.com/synthetics/beta">Datadog Synthetics request form</a>.</div>

Reviewer 3: Should we also add a link to the form for EU based accounts (https://app.datadoghq.eu/synthetics/beta) or is it useless?

Reviewer 4: Useless: we don't support EU orgs for now.

> <div class="alert alert-warning">Synthetics is in beta for the Datadog US Site. To request access, complete the <a href="https://app.datadoghq.com/synthetics/beta">Datadog Synthetics request form</a>.</div>

Reviewer 5: ```suggestion
<div class="alert alert-warning">Synthetics is in beta for the Datadog US Site. Request access via the <a href="https://app.datadoghq.com/synthetics/beta">Datadog Synthetics request form</a>.</div>
```

## Overview

> Browser Checks are scenarios executed by Datadog on your web application at configurable periodic intervals, from multiple locations around the world, on multiple devices. These checks verify not only that your applications are up and responding to requests, but that any conditions defined in your scenarios are met.

Reviewer 5: ```suggestion
Browser Checks are scenarios executed by Datadog on your web applications. They run at configurable periodic intervals from multiple locations around the world, and from multiple devices. These checks verify both that your applications are up and responding to requests, and that any conditions defined in your scenarios are met.
```

## Configuration

### Test details

> Define the configuration of our Browser Check.

Reviewer 2: ```suggestion
Define the configuration of your Browser Check.
```

{{< img src="synthetics/browser_check/browser_check_configuration.png" alt="Browser Check make request" responsive="true" style="width:80%;">}}

1. Give your Browser Check a name. 

> 2. Enter the Start URL. This is URL on which your Browser Check starts its scenario.

Reviewer 5: ```suggestion
2. Enter the Start URL. This is URL from which your Browser Check starts the scenario.
```

> 2. Enter the Start URL. This is URL from which your Browser Check starts the scenario.

Reviewer 2: ```suggestion
2. Enter the Start URL. This is the URL from which your Browser Check starts the scenario.
```

> 2. Enter the Start URL. This is the URL from which your Browser Check starts the scenario.

Reviewer 3: Are we not talking about tags on purpose? It might be worth adding a blurb on the ability to select tags that will allow filtering on the main Synthetics page afterwards?

Reviewer 4: Good idea!

3. Define on which device to run your check. Available devices are: `Laptop Large`, `Tablet`, and `Mobile Small`.

> 4. Pick-up locations to run the test from. Available locations are:

Reviewer 5: Is "pick-up" a noun or a verb here?

If noun (and only one is possible):
> Define a pick-up location to run the …

If noun and multiples are possible:
> Define one or more pick-up locations to run the …

If verb... uh that's a weird verb, but ok:
> Pick-up a location to run the …

Author: it's a verb, we could use `select`/`configure`.. is it less weird?

Reviewer 2: Go with select or configure.

    * Frankfurt (Request made from an AWS Datacenter)
    * Tokyo (Request made from an AWS Datacenter)
5. Choose a Check frequency between "1 run per 5 minute interval" to "1 run per week":

    {{< img src="synthetics/browser_check/check_frequency.png" alt="Check frequency" responsive="true" style="width:80%;">}}

### Notifications

A notification is sent if at least one step of the Browser Check scenario fails. To configure your notifications:

{{< img src="synthetics/browser_check/browser_check_notification.png" alt="Browser check notification" responsive="true" style="width:80%;">}}

1. Select users and/or [services][1] to send the notifications to.
2. Enter a **message** for the Browser Check. This field allows standard [Markdown formatting][2]. Notification messages include the **message** defined in this section and information about which assertion failed and why.
3. Click **Save Details and Record Test** to save your Browser Check.

> 4. Start to record your test.

Reviewer: Don't know if it worth explaining that, but the process is
- when a check is created for the first time, it's created *paused*
- the first time you click "Save Details and Record Test", you go to the test recorder page in a special mode where you can "Save And Launch" which will save the steps and un-pause the check
- the other time you go to the test recorder page, you can only "Save" and "Save And Quit" which save steps but won't change the status pause / live of the check.

Also something subtle is that, if a check is live but with zero steps, we'll still run it by adding ourself a single step which is the navigation to the start URL of the check. (This is indicated by a warning in the check details page).

Author: no need to complexify the doc here, the DD UI is clear enough about this.

## Record test

**Tests can be only recorded from [Google Chrome][3]. To record your test, download the [Datadog Record Test extension for Google Chrome][4].**

{{< img src="synthetics/browser_check/browser_check_record_test.png" alt="Browser check record test" responsive="true" >}}

> 1. Optionally, select **Open in a pop-up** at the upper right of the page to open your test recording in a separate pop-up window in order to avoid sizing issues in the displayed window within Datadog interface.

Reviewer 2: ```suggestion
1. Optionally, select **Open in a pop-up** at the upper right of the page to open your test recording in a separate pop-up window in order to avoid sizing issues in the displayed window within Datadog's interface.
```

> 1. Optionally, select **Open in a pop-up** at the upper right of the page to open your test recording in a separate pop-up window in order to avoid sizing issues in the displayed window within Datadog's interface.

Reviewer 3: > in order to avoid sizing issues in the displayed window within Datadog's interface.
Should we have that bit considering the dropdown that will soon (I believe) be added to the iframe to be able to select the device type? @GJSafar

Reviewer 4: This feature is coming soon.
But let's wait until it's here before we add it to the doc.

2. Click on **Start recording** to begin recording your Browser Check.
3. Your actions are recorded and used to create steps within your Browser Check Scenario.

> 4. Optionally, use the Actions available in the upper left corner to enrich your scenario:

Reviewer 2: ```suggestion
4. Optionally, use the actions available in the upper left corner to enrich your scenario:
```

>     {{< img src="synthetics/browser_check/browser_check_assertions.png" alt="Browser Check assertions" responsive="true" style="width:80%;">}}

Reviewer 3: Could we detail the actions a bit more? Especially the Variable action which is pretty useful (to fill forms for instance). [By clicking on the hand and then selecting a field on the page, we're able to assign it the variable OR by selecting Use Variable in the Type text action.](https://cl.ly/642ddd626e07)

Reviewer: Also there is nothing on the different types of assertions, hover and navigate steps, is that on purpose?

5. Once you have finished your Scenario, click on **Save and Launch Test**.

> 

Reviewer 3: Should we add a note on how it's relevant and recommended to finish the scenario by an assertion maybe? @GJSafar

Reviewer 4: Yes, it's very important: if one of the keys identifying the elements is broker while the others are not, the broken key is repaired if and only if the check completely succeeds. Hence, it's very important to make sure the success condition is very clear, to avoid having a "drift" of keys.

No need to explain it that way, but, we have to explain the check should succeed based on the same kind of assertions a human being should be asking. "Does my journey end-up on the right page? --> how do I know it's the right page? --> what text/value indicates me I'm on the right page? --> I create an assertion to validate that text/value are displayed as expected on the page"

Reviewer: Second that.

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: /integrations/#cat-notification
[2]: http://daringfireball.net/projects/markdown/syntax
[3]: https://www.google.com/chrome

> [4]: https://chrome.google.com/webstore/detail/datadog-test-recorder-sta/bfgpghinhlklmedgkkpnhphfgdliceel

Reviewer: This is the staging extension, which should be used only internally. We're having difficulties with the Chrome webstore so right now it's also used in production but this may change soon. I cannot give the production extension link now and it's not clear what will be our actions on the subject, so if possible I'd advise to not put the extension link here and just explain that the extension can be downloaded from the test recorder page.
